### ISSUE TITLE
Feature request: allow user-specified font strings

### ISSUE THREAD

**Issue by athompso** [primeau/Larn#4]
It's great that I can flip between dos437 and ... eh, whatever "Modern" was, I've already forgotten, but neither of them look good on my screen.  I can use my browser's developer tools to change the font, but it doesn't seem to persist through a game.
I would like to be able to just enter the font string directly.
Not sure what the best way to do this would be, and I can't write JS so am not able to try implementing it myself.

**Comment by primeau** [comment]
Aww, no love for the classic "Courier New"? :)

What's your preferred font? I'd be happy to give it a try and see how it looks.

In the meantime, your best bet would be to do a search-and-replace on "dos437" in the code (display.js and larn.css) with your font and re-build. I know you said you're not a JS dev but the build instructions in the readme should just work. I can't stand the JS ecosystem so I made this project as bare bones as possible. Let me know if you have any trouble and I'd be happy to help you out.

**Comment by athompso** [comment]
I will need to spin up a dedicated VM if I have to install `npm`, in order to keep it quarantined.  Oh, well, that's not impossible.

Part of my specific issue is that I'm running under Windows on a 4k monitor with scaling set to 125% so some fonts just render _strangely_, and on the 55" monitor in front of me it's really glaringly obvious when it happens.

The "Modern" mode (Courier New?) doesn't work at all for me - the "@" to indicate the player character vanishes!  I guess I should open a separate issue for that one of these days...

I've tried Hack, Consolas, the set of { Lucida Sans Typewriter, Bitstream Vera Sans Mono, DejaVu Sans Mono, Liberation Mono, Ubuntu Mono - they're all basically the same}, Droid Sans Mono, Lucida Console, Noto Mono, OCR A, Source Code Pro, and several others I don't currently have installed, for both Nethack and [U]Larn over the years.

As much as I hate to admit it, Microsoft's Consolas actually works quite well for roguelikes.  It's got far too short a height to be pleasant to read text, but its squarer aspect ratio works better than e.g. the Bitstream Vera or Lucida fonts which although otherwise pleasant, are too tall and skinny to render a dungeon very convincingly.  Hack works reasonably well, too.

(Yes, I'm a typography snob, in case that wasn't immediately obvious.)

**Comment by athompso** [comment]
Oh, if you want a **real** "retro" feel, I might suggest VT323 or ModeSeven, although I don't know if their glyph coverage will be adequate - both are pretty small.

**Comment by primeau** [comment]
Bummer, they actually both look *great* on the intro screen, but totally fall apart after that...

VT323 intro:
![image](https://user-images.githubusercontent.com/1916587/152477947-b54dc244-1094-4829-bb24-d2ec88b99db8.png)

ModeSeven intro:
![image](https://user-images.githubusercontent.com/1916587/152477786-85cde7b2-8739-42e8-a0cc-1575ce87612b.png)

VT323 maze:
![image](https://user-images.githubusercontent.com/1916587/152478109-aad06d4e-8b67-467a-81d8-d7f6f81e7a18.png)

ModeSeven maze:
![image](https://user-images.githubusercontent.com/1916587/152478193-31214105-06ee-4257-85bc-5f477364ca18.png)

Even with some tweaks VT323 is probably too narrow. I think ModeSeven might be fixed by making the '.' the same width as everything else, but I'm hopeless at editing fonts and have burned many, many hours failing to fix a few other fonts while I was adding the DOS and Amiga font options (Safari is still broken), so I have to bow out of trying again for my own sanity.

**Comment by athompso** [comment]
Wait, you're _editing fonts_???  Holy crap, dude.  But what you're seeing is also very strange...  Ever thought about re-implementing [n]curses in JS instead? :-D

**Comment by athompso** [comment]
Well, here's **[Hack](https://sourcefoundry.org/hack/)**, which looks OK:
![image](https://user-images.githubusercontent.com/887417/152479664-dd1fd64d-4f19-4e58-9771-8981d7b330a1.png)

**Comment by primeau** [comment]
Well, I *tried* to edit the DOS font so it would look 100% correct on Safari, but failed after many hours of effort so I gave up...

The problem you're seeing is that the dot char isn't actually monospaced so the rows are getting compressed. There's not much I can do about that without a bunch of effort to change the rendering engine, or successfully editing the effing fonts :)

I have to say I actually prefer the non-serif look of the Hack font over Courier in Modern mode, but the dots are a bit big for my liking. Have you got any other suggestions for non-serif monospaced fonts?

**Comment by athompso** [comment]
So far, I have a short-list of fonts that are known to be Monospaced, Sans-Serif, and have at least the wall and dot glyphs.  The vast majority of sans-serif monospace fonts are all just about clones of each other (and in many cases they are literally based off the earlier font).
I haven't gone through the list and tested them all on dungeon level 1, yet.  And as you can see it's not a fully-complete dataset, either.

Now I need to dive down a _different_ rabbit hole for a while, but I'll get back to this - it involves _FONTS_!

| Name                         | License | Serif | Designer                     | Foundry         | Distributor   | Family      | Metric-compatible with       | Availability                                                                                                         | Supports LARN? |
| ---------------------------- | ------- | ----- | ---------------------------- | --------------- | ------------- | ----------- | ---------------------------- | -------------------------------------------------------------------------------------------------------------------- | -------------- |
| Actual Mono Light            |         | Sans  |                              | Frutitype       |               |             |                              | [https://www.myfonts.com/fonts/frutitype/actual-mono/](https://www.myfonts.com/fonts/frutitype/actual-mono/)         | Y              |
| Code New Roman               |         | Sans  | Sam Radian                   |                 |               |             |                              |                                                                                                                      | Y              |
| Cousine                      |         |       |                              | Ascender        | Google        | Croscore    | Courier New                  | Chrome OS                                                                                                            | Y              |
| DejaVu Sans Mono             |         |       | \-                           | \-              | \-            | Deja Vu     | Bitstream Vera Sans Mono     | [https://dejavu-fonts.github.io/](https://dejavu-fonts.github.io/)                                                   | Y              |
| Fantasque Sans Mono          |         | Sans  | Jany Belluz                  | \-              | \-            | \-          | \-                           | [https://github.com/belluzj/fantasque-sans](https://github.com/belluzj/fantasque-sans)                               | Y              |
| Fira Mono                    |         | Sans  | Erik Spiekermann et al.      | Monotype        | bBox Type     | Fira        | \-                           | [https://bboxtype.com/typefaces/FiraMono](https://bboxtype.com/typefaces/FiraMono)                                   | Y              |
| Hack                         |         | Sans  |                              | SourceFoundry   | SourceFoundry | \-          |                              | [https://sourcefoundry.org/hack/](https://sourcefoundry.org/hack/)                                                   | Y              |
| Inconsolata                  |         | Sans  | Raph Levien                  | \-              | \-            | \-          | Consolas                     | [https://levien.com/type/myfonts/inconsolata.html](https://levien.com/type/myfonts/inconsolata.html)                 | Y              |
| JetBrains Mono               |         | Sans  |                              |                 | JetBrains     | \-          | \-                           | [https://www.jetbrains.com/lp/mono/](https://www.jetbrains.com/lp/mono/)                                             | Y              |
| Liberation Mono              |         | Sans  | Steve Matteson               | Ascender        | Red Hat       | Liberation  | Croscore/Cousine/Courier New | [https://github.com/liberationfonts](https://github.com/liberationfonts)                                             | Y              |
| Monofur                      |         | Sans  | Tobias Kohler (unci narynin) |                 |               | euroference |                              | [http://luc.devroye.org/fonts-24320.html](http://luc.devroye.org/fonts-24320.html)                                   | Y              |
| Monospaceland Extra Light    |         | Sans  |                              |                 |               |             |                              | [https://www.myfonts.com/fonts/pepper-type/monospaceland/](https://www.myfonts.com/fonts/pepper-type/monospaceland/) | Y              |
| NK57 Monospace               |         | Sans  |                              | Typpodermic     |               |             |                              |                                                                                                                      | Y              |
| NK57 Monospace Semicondensed |         | Sans  |                              | Typpodermic     |               |             |                              |                                                                                                                      | Y              |
| Noto Sans Mono               |         | Demi  |                              | Google          | Google        | Noto        |                              | [https://fonts.google.com/noto](https://fonts.google.com/noto)                                                       | Y              |
| Pexico Regular               |         | Sans  | Ondrej Job                   | Setup Type/URTD | MyFonts       |             |                              | [https://www.myfonts.com/fonts/urtd/pexico/](https://www.myfonts.com/fonts/urtd/pexico/)                             | Y              |
| ProFont                      | ?       |       | Tobias Jung et al.           |                 |               |             | Monaco                       | [https://tobiasjung.name/profont/](https://tobiasjung.name/profont/)                                                 | Y              |
| PT Mono                      | SIL OFL | Sans  | Alexandra Korolkova          | ParaType        | Rospechat     | Public Type |                              | [https://www.paratype.com/fonts/pt/pt-mono](https://www.paratype.com/fonts/pt/pt-mono)                               | Y              |
| Source Code Pro              | SIL OFL | Sans  | Paul D. Hunt                 | Adobe           |               |             |                              | [https://fonts.adobe.com/fonts/source-code-pro](https://fonts.adobe.com/fonts/source-code-pro)                       | Y              |
| Ubuntu Mono                  |         |       | Dalton Maag Ltd.             |                 | Canonical     |             |                              |                                                                                                                      | Y              |
| Fira Code                    |         | Sans  | Erik Spiekermann et al.      | Monotype        | bBox Type     | Fira        | \-                           | [https://github.com/bBoxType/FiraSans](https://github.com/bBoxType/FiraSans)                                         | Y?             |

**Comment by athompso** [comment]
Oh, also, do you need a font in WOFF format?  I can't tell for sure here.

**Comment by primeau** [comment]
Check out the latest beta -- I went down my own Google Font rabbit hole a few days ago and thought Noto Sans looked really good so I went ahead and changed the modern font to that on Chrome. PT-Mono also looked good too. Thanks for the inspiration!

I still need to make something other than Courier work on Firefox and Safari, but that's a project for another day. Including the problematic characters would have saved me a ton of time!
